File: tools/asm_to_zig_prototype.py
import re
import sys

def translate(asm_path, zig_path):
    with open(asm_path, 'r') as f:
        lines = f.readlines()

    out = []
    out.append('const std = @import("std");')
    out.append('const reg_map = @import("register_mapping.zig");')
    out.append('const Register = reg_map.Register;')
    out.append('const Executor = @import("instruction_operations.zig").Executor;')
    out.append('const abi = @import("abi_handshake.zig");')
    out.append('')
    out.append('pub export fn rosetta3_run_snax86() void {')
    out.append('    const allocator = std.heap.page_allocator;')
    out.append('    var ex = Executor.init(allocator, 1024 * 1024);')
    out.append('    defer ex.deinit();')
    out.append('    ex.regs.esp = 0x100000; // 1MB stack top')
    out.append('    abi.register_snake_thunks(&ex);')
    out.append('')
    
    # Simple data section handling
    data_addr = 0x2000
    equs = {}
    
    current_section = None
    
    for line in lines:
        line = line.strip()
        if not line or line.startswith(';'): continue
        
        # Handle section
        m = re.match(r'section\s+\.(data|text)', line)
        if m:
            current_section = m.group(1)
            continue
            
        # Handle EQU
        m = re.match(r'(\w+)\s+equ\s+(.*)', line)
        if m:
            name, val = m.group(1), m.group(2)
            # Try to evaluate simple math
            try:
                val = eval(val.replace('H', '0x').replace('h', '0x').replace('$', '0')) # crude
            except:
                pass
            equs[name] = val
            continue

        if current_section == 'data':
            # Handle db/dw/dd
            m = re.match(r'(\w+)\s+(db|dw|dd|times)\s+(.*)', line)
            if m:
                name, size, val = m.group(1), m.group(2), m.group(3)
                out.append(f'    // Data: {name} at 0x{data_addr:x}')
                out.append(f'    ex.labels.put("{name}", 0x{data_addr:x}) catch unreachable;')
                if size == 'db':
                    if val.startswith('"'):
                        s = val.strip('"')
                        for char in s:
                            out.append(f'    ex.mem.write8(0x{data_addr:x}, {ord(char)});')
                            data_addr += 1
                    else:
                        out.append(f'    ex.mem.write8(0x{data_addr:x}, {val});')
                        data_addr += 1
                elif size == 'dw':
                    out.append(f'    ex.mem.write16(0x{data_addr:x}, {val});')
                    data_addr += 2
                elif size == 'dd':
                    out.append(f'    ex.mem.write32(0x{data_addr:x}, {val});')
                    data_addr += 4
                continue

    out.append('')
    out.append('    // Entry point')
    out.append('    _main(&ex);')
    out.append('}')
    out.append('')

    # Translate functions
    current_func = None
    for line in lines:
        line = line.split(';')[0].strip()
        if not line: continue
        
        # Labels
        m = re.match(r'^(\w+):', line)
        if m:
            name = m.group(1)
            if current_func:
                out.append('}')
            current_func = name
            out.append(f'fn {name}(ex: *Executor) void {{')
            continue
        
        if not current_func: continue
        
        # Instructions (very crude mapping)
        parts = re.split(r'[\s,]+', line)
        inst = parts[0].lower()
        args = parts[1:]
        
        if inst == 'call':
            target = args[0]
            if target.startswith('_'):
                out.append(f'    ex.call_external("{target}");')
            else:
                out.append(f'    {target}(ex);')
        elif inst == 'mov':
            dst, src = args[0], args[1]
            if dst in ['eax', 'ebx', 'ecx', 'edx', 'esi', 'edi', 'ebp', 'esp']:
                if src.isdigit() or src.startswith('0x'):
                    out.append(f'    ex.mov_reg_imm(.{dst}, {src});')
                elif src in ['eax', 'ebx', 'ecx', 'edx', 'esi', 'edi', 'ebp', 'esp']:
                    out.append(f'    ex.mov_reg_reg(.{dst}, .{src});')
        elif inst == 'ret':
            out.append('    return;')

    if current_func:
        out.append('}')

    with open(zig_path, 'w') as f:
        f.write('\n'.join(out))

if __name__ == '__main__':
    translate(sys.argv[1], sys.argv[2])

File: tools/asm_to_zig_prototype.zig
const std = @import("std");

const ParsedLine = union(enum) {
    section: []const u8,
    equ: struct { name: []const u8, value: []const u8 },
    label: []const u8,
    instruction: struct { op: []const u8, args: []const u8 },
    data_decl: struct { name: []const u8, kind: []const u8, value: []const u8 },
    ignore,
};

fn trimComment(line: []const u8) []const u8 {
    var split = std.mem.splitScalar(u8, line, ';');
    return std.mem.trim(u8, split.first(), " \t\r");
}

fn parseLine(line: []const u8) ParsedLine {
    const trimmed = trimComment(line);
    if (trimmed.len == 0) return .ignore;

    if (std.mem.startsWith(u8, trimmed, "section .")) {
        return .{ .section = trimmed["section .".len..] };
    }

    if (std.mem.indexOf(u8, trimmed, " equ ")) |idx| {
        return .{ .equ = .{
            .name = std.mem.trim(u8, trimmed[0..idx], " \t"),
            .value = std.mem.trim(u8, trimmed[idx + 5 ..], " \t"),
        } };
    }

    if (trimmed[trimmed.len - 1] == ':') {
        return .{ .label = trimmed[0 .. trimmed.len - 1] };
    }

    const data_kinds = [_][]const u8{ " db ", " dw ", " dd ", " times " };
    for (data_kinds) |kind| {
        if (std.mem.indexOf(u8, trimmed, kind)) |idx| {
            return .{ .data_decl = .{
                .name = std.mem.trim(u8, trimmed[0..idx], " \t"),
                .kind = std.mem.trim(u8, kind, " \t"),
                .value = std.mem.trim(u8, trimmed[idx + kind.len ..], " \t"),
            } };
        }
    }

    const first_space = std.mem.indexOfScalar(u8, trimmed, ' ') orelse trimmed.len;
    return .{ .instruction = .{
        .op = trimmed[0..first_space],
        .args = if (first_space < trimmed.len) std.mem.trim(u8, trimmed[first_space + 1 ..], " \t") else "",
    } };
}

fn writeHeader(writer: anytype) !void {
    try writer.writeAll(
        \\const std = @import("std");
        \\const Executor = @import("instruction_operations.zig").Executor;
        \\const abi = @import("abi_handshake.zig");
        \\
        \\pub export fn rosette_run_prototype() void {
        \\    const allocator = std.heap.page_allocator;
        \\    var ex = Executor.init(allocator, 1024 * 1024);
        \\    defer ex.deinit();
        \\    ex.regs.esp = 0x100000;
        \\
    );
}

fn writeFooter(writer: anytype) !void {
    try writer.writeAll(
        \\
        \\    _main(&ex);
        \\}
        \\
    );
}

fn translateInstruction(writer: anytype, op: []const u8, args: []const u8) !void {
    if (std.ascii.eqlIgnoreCase(op, "ret")) {
        try writer.writeAll("    return;\n");
        return;
    }

    if (std.ascii.eqlIgnoreCase(op, "call")) {
        if (args.len > 0 and args[0] == '_') {
            try writer.print("    // external call {s}\n", .{args});
        } else {
            try writer.print("    {s}(ex);\n", .{args});
        }
        return;
    }

    if (std.ascii.eqlIgnoreCase(op, "mov")) {
        var split = std.mem.splitScalar(u8, args, ',');
        const dst = std.mem.trim(u8, split.next() orelse "", " \t");
        const src = std.mem.trim(u8, split.next() orelse "", " \t");
        const regs = [_][]const u8{ "eax", "ebx", "ecx", "edx", "esi", "edi", "ebp", "esp" };

        var dst_reg = false;
        var src_reg = false;
        for (regs) |reg| {
            if (std.ascii.eqlIgnoreCase(dst, reg)) dst_reg = true;
            if (std.ascii.eqlIgnoreCase(src, reg)) src_reg = true;
        }

        if (dst_reg and src_reg) {
            try writer.print("    ex.mov_reg_reg(.{s}, .{s});\n", .{ dst, src });
            return;
        }
        if (dst_reg) {
            try writer.print("    ex.mov_reg_imm(.{s}, {s});\n", .{ dst, src });
            return;
        }
    }

    try writer.print("    // TODO: {s} {s}\n", .{ op, args });
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.arena.allocator();
    const args = try init.minimal.args.toSlice(allocator);
    if (args.len != 3) {
        std.debug.print("usage: {s} <input.asm> <output.zig>\n", .{args[0]});
        return error.InvalidArguments;
    }

    const asm_source = try std.Io.Dir.cwd().readFileAlloc(init.io, args[1], allocator, .limited(1024 * 1024));
    var output: std.ArrayList(u8) = .empty;
    var writer_state: std.Io.Writer.Allocating = .fromArrayList(allocator, &output);
    defer output = writer_state.toArrayList();
    const writer = &writer_state.writer;

    try writeHeader(writer);

    var current_section: ?[]const u8 = null;
    var line_iter = std.mem.splitScalar(u8, asm_source, '\n');
    while (line_iter.next()) |line| {
        switch (parseLine(line)) {
            .ignore => {},
            .section => |name| current_section = name,
            .equ => |entry| try writer.print("    // EQU {s} = {s}\n", .{ entry.name, entry.value }),
            .data_decl => |decl| {
                if (current_section != null and std.mem.eql(u8, current_section.?, "data")) {
                    try writer.print("    // DATA {s} {s} {s}\n", .{ decl.name, decl.kind, decl.value });
                }
            },
            .label => |name| try writer.print("fn {s}(ex: *Executor) void {{\n", .{name}),
            .instruction => |inst| try translateInstruction(writer, inst.op, inst.args),
        }
    }

    try writeFooter(writer);
    try std.Io.Dir.cwd().writeFile(init.io, .{ .sub_path = args[2], .data = output.items });
}
